README.md
# theproject

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar theproject-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2018 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-02-08
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-02-08
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/theproject/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/theproject/compare/0.1.0...0.1.1


test/theproject/core_test.clj
(ns theproject.core-test
  (:require [clojure.test :refer :all]
            [theproject.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/theproject/core.clj
(ns theproject.core
  (:gen-class)
  (:require [clojure.data.json :as json])
  (:require [cheshire.core :refer :all])
  (:require [datomic.api :as d]))

(def data_output {:job_assigned {:job_id "job id" :agent_id "agent id"}})

(defn read-json-data [archivepath] 
  (parse-string (slurp archivepath) true))

(def archivepath "resource/sample-input.json")

(def datareceived (read-json-data archivepath))

(defn get-agents [coll] (filter #(:new_agent %) coll))

(defn get-jobs [coll] (filter #(:new_job %) coll))

(defn get-jobs-requested [coll] (filter #(:job_request %) coll))

(defn get-urgent-jobs [coll]
  (filter #(= (:urgent (:new_job %)) true) coll))

(defn get-noturgent-jobs [coll] 
  (filter #(= (:urgent (:new_job %)) false) coll))

;;Don't use concat https://stuartsierra.com/2015/04/26/clojure-donts-concat
(defn prepare-coll [urgent_coll noturgent_coll] 
  (into urgent_coll noturgent_coll))

(defn select-available-agent [coll_jobs_requested coll_agents agents_list] 
  (cond 
    (or (empty? coll_jobs_requested) (empty? coll_agents)) 
      agents_list
    (= (:agent_id (:job_request (first coll_jobs_requested)) (:id (:new_agent (first coll_agents))))) 
      (recur (rest coll_jobs_requested) (rest coll_agents) (conj agents_list (first coll_agents))) 
    :else 
      (recur coll_jobs_requested (rest coll_agents) agents_list)))

(defn check-skill [skill_list skill] 
  (not (nil? (some #{skill} skill_list))))

(defn select-primally-skill [coll_jobs coll_agents] 
  ())

  
; (println (check-skill ["asdf" "jklç" "qwer" "poiu"] "jklç"))

; (println (select-available-agent (get-jobs-requested datareceived) (get-agents datareceived) '()))

; (defn get-available-agents [coll_agents :job_request])
  
; (defn availa)

; (defn select-agent [coll_agent coll_job_request job] 
;   (loop []))


; (defn dequeue [coll]
;  (let  [output ()]
;   ( )))

;(println (get-urgent-jobs (get-jobs datareceived)))
;(println (get-noturgent-jobs (get-jobs datareceived)))


(defn -main
  "I don't do a whole lot ... yet."
  [& args] 

  ; (let [in (slurp *in*)]
  ;   (println (parse-string in)))

  ;; Reciving data, done!
  ;(def data-readed (parse-string (slurp *in*)))
)




src/theproject/outline/outline.clj
(ns theproject.core
    (:gen-class)
    (:require [clojure.data.json :as json])
    (:require [cheshire.core :refer :all])
    (:require [datomic.api :as d])
  )
  
  ;; to run:  cat resource/sample-input.json | lein trampoline run
  
  ;;hard way
  ;;  (use 'clojure.java.io)
  ;;  (with-open [wrtr (writer "/tmp/test.json")]
  ;;     (.write wrtr (json/write-str {:key1 "val1" :key2 "val2"})))
  ;;lein dep && lein compile
  
  ;;> (spit "/tmp/test.json" (json/write-str {:key1 "val1" :key2 "val2"}))
  
  ;;(defn read-json [path] (json/read-str (read-archive path)) :value-fn keyword)
  ;;(map #(cheshire/parse-string % true)
  ;;(line-seq (clojure.java.io/reader "data/json_data")))
  
  
  ;; Data tratament
  
  ;(defn read-archive [path] (slurp path))
  
  
  (def data_output {:job_assigned {:job_id "job id" :agent_id "agent id"}})
  
  
  (defn read-json-data [archivepath] 
    (parse-string (slurp archivepath) true))
  
  (def archivepath "resource/sample-input.json")
  
  (def datareceived (read-json-data archivepath))
  
  ;;(map #(parse-string % true)
  ;;     (line-seq (clojure.java.io/reader "/Users/josevictorpereiracosta/Documents/secret/jobqueue/resource/jason_data.json")))
  
  ;;(defn teste [archivepath] (parse-string (slurp archivepath) true))
  
  ; ;; With fn
  ; (defn get-agents [coll] (filter (fn [x] (:new_agent x)) coll))
  
  ; (defn get-jobs [coll] (filter (fn [x] (:new_job x)) coll))
  
  ; (defn get-jobs-request [coll] (filter (fn [x] (:job_request x)) coll))
  
  ;;With high order function
  
  
  
  
  ;;With if
  ; (defn get-urgent-jobs [coll] 
  ;   (filter (fn [x] (not= nil x)) (map is-urgent coll)))
  
  ;;Without if
  
  
  (defn get-agents [coll] (filter #(:new_agent %) coll))
  
  (defn get-jobs [coll] (filter #(:new_job %) coll))
  
  (defn get-jobs-request [coll] (filter #(:job_request %) coll))
  
  ; (defn is-urgent [coll] 
  ;   (= ((coll :new_job) :urgent) true))
  
  ; (defn is-not-urgent [coll] 
  ;   (= ((coll :new_job) :urgent) false))
  
  (defn get-urgent-jobs [coll]
    (filter #(= (:urgent (:new_job %)) true) coll))
  
  (defn get-noturgent-jobs [coll] 
    (filter #(= (:urgent (:new_job %)) false) coll))
  
  
  ;;Don't use concat https://stuartsierra.com/2015/04/26/clojure-donts-concat
  (defn prepare-coll [urgent_coll noturgent_coll] 
    (into urgent_coll noturgent_coll))
  
  
    
  (defn dequeue [coll]
   (let  [output ()]
    ( )))
  
  ;(println (get-urgent-jobs (get-jobs datareceived)))
  ;(println (get-noturgent-jobs (get-jobs datareceived)))
  (println (prepare-coll (get-noturgent-jobs (get-jobs datareceived)) (get-urgent-jobs (get-jobs datareceived))))
  
  ;  (defn dequeue [coll] 
  ;    (let [])
  
  (defn -main
    "I don't do a whole lot ... yet."
    [& args] 
  
    ; (let [in (slurp *in*)]
    ;   (println (parse-string in)))
  
    ;; Reciving data, done!
    ;(def data-readed (parse-string (slurp *in*)))
  )
  
  
  